File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        global root
        root = None
        def construct(preorder, inorder):
            global root
            if not preorder and not inorder:
                return None            
            head = TreeNode(preorder[0])
            
            if not root:
                root = head
            
            i = 0
            n_inorder = len(inorder)
            while(i < n_inorder and inorder[i] != head.val):
                i = i + 1
            left_inorder = inorder[:i]
            right_inorder = inorder[i+1:]
            n_left_inorder = len(left_inorder)
            n_right_inorder = len(right_inorder)
            
            left_preorder = preorder[1:n_left_inorder+1]
            right_preorder = preorder[n_left_inorder+1:]
            
            left_node = construct(left_preorder, left_inorder)
            right_node = construct(right_preorder, right_inorder)
            head.left = left_node
            head.right = right_node
            return head
        construct(preorder, inorder)
        return root
            
            
            
        

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int preLen = preorder.length, inLen = inorder.length;
        if(preLen == 0){
            return null;
        }
        if(preLen == 1 && inLen == 1){
            return new TreeNode(preorder[0]);
        }
        TreeNode root = new TreeNode(preorder[0]);
        
        //find the root in inorder
        int i = 0;
        for(i = 0; i < inLen; i++){
            if(root.val == inorder[i]){
                break;
            }
        }
        TreeNode leftSubTree = buildTree(Arrays.copyOfRange(preorder,1, i+1), Arrays.copyOfRange(inorder, 0, i) );
        TreeNode rightSubTree = buildTree(Arrays.copyOfRange(preorder,i+1, preLen), Arrays.copyOfRange(inorder, i+1, inLen) );
        root.left = leftSubTree;
        root.right = rightSubTree;
        return root;
    }
}